validations/src/validations.rs

```diff
@@ -215,63 +237,109 @@ fn update_liars(liars: &mut Vec<bool>, item: RadonTypes, condition: bool) -> Opt
     }
 }
 
+/// An `Either`-like structure that covers the two possible return types of the
+/// `evaluate_tally_precondition_clause` method.
+#[derive(Debug)]
+pub enum TallyPreconditionClauseResult {
+    MajorityOfValues {
+        values: Vec<RadonTypes>,
+        liars: Vec<bool>,
+    },
+    MajorityOfErrors {
+        errors_mode: RadonError<RadError>,
+    },
+}
+
+/// Run a precondition clause on an array of `RadonTypes` so as to check if the mode is a value or
+/// an error, which has clear consequences in regards to consensus, rewards and punishments.
 pub fn evaluate_tally_precondition_clause(
     reveals: Vec<RadonReport<RadonTypes>>,
-    non_error_min: f64,
-) -> Result<(Vec<RadonTypes>, Vec<bool>), RadError> {
+    minimum_consensus: f64,
+) -> Result<TallyPreconditionClauseResult, RadError> {
+    // Short-circuit if there were no reveals
     if reveals.is_empty() {
         return Err(RadError::NoReveals);
     }
 
-    let reveals_len = reveals.len() as f64;
-
+    // Count how many times is each RADON type featured in `reveals`, but count `RadonError` items
+    // separately as they need to be handled differently.
+    let reveals_len = reveals.len() as u32;
     let mut counter = Counter::new(RadonTypes::num_types());
     for reveal in &reveals {
-        match &reveal.result {
-            Err(_e) => {}
-            Ok(rad_types) => counter.increment(rad_types.discriminant()),
-        }
-    }
-
-    let non_error_ratio = counter.max_val as f64 / reveals_len;
-
-    if non_error_ratio > non_error_min {
-        let mut liars = vec![];
-        let results = reveals
-            .into_iter()
-            .filter_map(|reveal| match reveal.into_inner() {
-                Err(_e) => {
-                    liars.push(true);
-                    None
+        counter.increment(reveal.result.discriminant());
+    }
+
+    // Compute ratio of type consensus amongst reveals (percentage of reveals that have same type
+    // as the frequent type).
+    let achieved_consensus = counter.max_val as f64 / reveals_len as f64;
+
+    // If the achieved consensus is over the user-defined threshold, continue.
+    // Otherwise, return `RadError::InsufficientConsensus`.
+    if achieved_consensus > minimum_consensus {
```

Comment:
If `minimum_consensus` is set to `1.0`, this branch can never be entered.

```suggestion
    if achieved_consensus >= minimum_consensus {
```

Reply:
Add tests

Reply:
Specifically for the `==` case?